docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,368 @@
+
+Given that there are other project in SIG cluster lifecycle, like e.g. https://github.com/kubernetes-sigs/cluster-addons , as well as well established solutions like helm, kapp, ArgoCD, flux, we should consider if/how to integrate Cluster API with those solutions.
+
+While considering this we should look at the addon lifecycle management in a holistic way and cover all the following requirements and possibly more:
+
+- Addon should inherit some configurations from cluster configuration (e.g service or pod CIDR for CNI addons, or the cluster name to be used as LB name in Azure CCM).
+- Addon should be created during cluster creation (e.g immediately after the API server is installed); this operation usually brings in some additional requirements, like: 
+  - Determine which addon must be installed for each Cluster.
+  - Resolve the addon version for the Cluster’s Kubernetes version.
+  - Provide addon configuration starting from a set of default settings but allowing per Cluster customization.
+  - Wait for the addon to be actually running.
+- Some of the addon must be upgraded before/during/after cluster (and more specifically control plane) upgrade; also in this case this operation brings in some additional requirements, like e.g. to proper cleanup the previous version of the addon.
+- There should be support for out-of band upgrades, e.g. for CVE fix (not linked to the cluster lifecycle).
+- It should be possible to orchestrate pre/post installation steps for each addon, thus allowing to manage migrations/operations like:
+  - CPI migrating from in-tree to out-of-tree
+  - CoreDNS configuration migration from one version to another
+- Addon deletion should be orchestrated before Cluster deletion
+
+The current proposal starts to address part of this problem by designing how Cluster API can orchestrate an external tool for addon management; the external tool is expected to provide all the foundational capabilities for addon management like e.g. addon packages/repository, templating/configuration, management of the creation, upgrade and deletion workflow etc.
+
+Cluster API - and most specifically a new Cluster API-aware component described in this proposal - should orchestrate the above tool so that Cluster lifecycle and addon lifecycle are consistent.
+
+
+### Goals
+
+- To design a solution for orchestrating Cluster add-on
+- To leverage an existing package management tool, Helm in this case, for all the foundational capabilities of addon management like e.g. addon packages/repository, templating/configuration, addon creation, upgrade and deletion etc. 
```

Comment:
```suggestion
- To leverage existing package management tools, for all the foundational capabilities of addon management like e.g. addon packages/repository, templating/configuration, addon creation, upgrade and deletion etc. 
```

Reply:
- To make addon management in CAPI modular and pluggable and to make it easy for anyone to build a CAPI addon provider based on their favorite package management tool, just like infrastructure and bootstrap providers.